pkg/config/store/fsstore2.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha1"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/golang/glog"
+)
+
+const defaultDuration = time.Second / 2
+
+type resourceMeta struct {
+	Name      string
+	Namespace string
+}
+
+type resource struct {
+	Kind       string
+	APIVersion string `json:"apiVersion"`
+	Metadata   resourceMeta
+	Spec       map[string]interface{}
+	sha        [sha1.Size]byte
+}
+
+func (r *resource) Key() Key {
+	return Key{Kind: r.Kind, Namespace: r.Metadata.Namespace, Name: r.Metadata.Name}
+}
+
+// fsStore2 is Store2Backend implementation using filesystem.
+type fsStore2 struct {
+	root          string
+	kinds         map[string]bool
+	checkDuration time.Duration
+	watchCtx      context.Context
+	watchCh       chan BackendEvent
+
+	mu   sync.RWMutex
+	data map[Key]*resource
+}
+
+var _ Store2Backend = &fsStore2{}
+
```

Comment:
Add comment that path is only used for error reporting.

Reply:
done